set bottom toolbar to show by default resolve issue #973

jhbiggs opened the pull request:
Fixes #973 

Changes: I set the default bottom toolbar to "enabled" in the preferences activity and the XML layout.

Screenshots/GIF for the change:
![Screen Shot 2019-06-10 at 10 22 13 PM](https://user-images.githubusercontent.com/35543657/59241691-4a71b300-8bce-11e9-8c37-4781b9d18fbd.png)
![Screen Shot 2019-06-10 at 10 22 23 PM](https://user-images.githubusercontent.com/35543657/59241705-58bfcf00-8bce-11e9-8822-896621efbc40.png)

jhbiggs commented:
Would you like to remove the option to disable/enable it from the Shared Preferences?

mhutti1 commented:
Yes

jhbiggs commented:
I took out the option in settings to show/hide the bottom toolbar.

mhutti1 commented:
This is failing travis because one of the tests we have involves clicking the button. You will need to fix this.

jhbiggs commented:
I was able to see that an error was on line 406 of the gradle.build file, but I didn't change this file in my commit.  I couldn't see any other problem.  I must be missing something.  Is there an artifacts page like there is with CircleCI?

Which button is being clicked?  I'm sorry, I don't understand.

jhbiggs commented:
Ok, I guessed that the check failed because the Shared Preferences expected a toggle switch in the settings where there was none.  So I added it back into the xml resource, but I disabled it.  I tried to test it out with TravisCI on my own before committing here, but I think I needed an API key which I don't have.

macgills commented:
@jhbiggs delete the preference in code and xml completely and with a device connected run `gradlew connectedKiwixDebugAndroidTest`, this will run the instrumentation tests. I haven't looked at the test in question but you will most likely have to delete a test or just a portion of a test. Once you have found the failing test you can run that test exclusively by placing your cursor inside the test method and pressing ctrl+shift+f10

jhbiggs commented:
@macgills When can I run a test exclusively?  How do I place a cursor in the right method?  The html output from the gradlew connectedKiwixDebugAndroidTest  showed INSTALL_FAILED_INSUFFICIENT_STORAGE.  I'm lost, sorry everyone.

I eliminated the bottom toolbar as requested.

macgills commented:
@jhbiggs are you running on an emulator or real device? Make sure kiwix is uninstalled before you start the tests. Our instrumentation test classes are located in the `app/src/androidTest*` folders, each of these test classes that has an `@Test` annotated method is a test, to expedite this I will switch to your branch now and run the tests.

macgills commented:
```
androidx.test.espresso.PerformException: Error performing 'load adapter data' on view 'is assignable from class: class android.widget.AdapterView'.
at androidx.test.espresso.PerformException$Builder.build(PerformException.java:84)
at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:85)
at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:57)
at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:316)
at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:177)
at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:118)
at androidx.test.espresso.DataInteraction$DisplayDataMatcher$1.apply(DataInteraction.java:213)
at androidx.test.espresso.DataInteraction$DisplayDataMatcher$1.apply(DataInteraction.java:210)
at androidx.test.espresso.DataInteraction$DisplayDataMatcher.<init>(DataInteraction.java:230)
at androidx.test.espresso.DataInteraction$DisplayDataMatcher.<init>(DataInteraction.java:205)
at androidx.test.espresso.DataInteraction$DisplayDataMatcher.displayDataMatcher(DataInteraction.java:248)
at androidx.test.espresso.DataInteraction.makeTargetMatcher(DataInteraction.java:150)
at androidx.test.espresso.DataInteraction.perform(DataInteraction.java:135)
at org.kiwix.kiwixmobile.tests.SettingsTest.testToggle(SettingsTest.java:55)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at androidx.test.runner.AndroidJUnit4.run(AndroidJUnit4.java:104)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:388)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2075)
Caused by: java.lang.RuntimeException: No data found matching: (is an instance of android.preference.Preference and  preference with key matching: is "pref_bottomtoolbar") contained values: <[Data: Display (class: android.preference.PreferenceCategory) token: 0, Data: Night mode Show articles with inverted colors (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 1, Data: Automated night mode Automatically switch between day and night mode. (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 2, Data: Back to Top Display a button at the end of the page to scroll up to the top (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 3, Data: Keep Toolbar Visible Keeps the toolbar from scrolling out of view when browsing content (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 4, Data: Extras (class: android.preference.PreferenceCategory) token: 5, Data: Open new tab in background When opening a new tab it will open in background (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 6, Data: Warn when entering external links Display popup to warn about additional costs or not working in offline links. (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 7, Data: Download content only via WiFi Download content only via WiFi (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 8, Data: Storage (class: android.preference.PreferenceCategory) token: 9, Data: Internal 22.1 GB (class: android.preference.Preference) token: 10, Data: Zoom (class: android.preference.PreferenceCategory) token: 11, Data: Zoom controls (class: org.kiwix.kiwixmobile.settings.CustomSwitchPreference) token: 12, Data: Zoom level medium (class: org.kiwix.kiwixmobile.settings.SliderPreference) token: 13, Data: History (class: android.preference.PreferenceCategory) token: 14, Data: Clear history Clear recent searches and tabs history (class: android.preference.Preference) token: 15, Data: Language (class: android.preference.PreferenceCategory) token: 16, Data: English Choose a language (class: android.preference.ListPreference) token: 17, Data: Information (class: android.preference.PreferenceCategory) token: 18, Data: Version 2.4 Build: 55 (class: android.preference.EditTextPreference) token: 19, Data: Credits Contributors and Licenses (class: android.preference.Preference) token: 20]>
at androidx.test.espresso.action.AdapterDataLoaderAction.perform(AdapterDataLoaderAction.java:95)
at androidx.test.espresso.ViewInteraction$SingleExecutionViewAction.perform(ViewInteraction.java:360)
at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:251)
at androidx.test.espresso.ViewInteraction.access$100(ViewInteraction.java:64)
at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:157)
at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:154)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:790)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6494)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```
SettingsTest.testToggle() is the failing method.

What you have to do is delete
```    onData(allOf(
        is(instanceOf(Preference.class)),
        withKey("pref_bottomtoolbar")))
        .perform(click());```

jhbiggs commented:
Hi,

I went back in and deleted the test above.  It still is failing.  There was a LanguageActivityTest  @Test that failed among others, but each time I commented out the test, another test failed.  I don't want to keep chasing my tail here unless you think I should.  I also deleted tests for HelpActivityTest and NetworkTest when they popped up. 

I don't have an Android device.  I have been using the emulator.  I will try to find a device here so I can run tests with more fidelity.

macgills commented on app/src/androidTestKiwix/java/org/kiwix/kiwixmobile/tests/SettingsTest.java:
```diff
@@ -49,10 +49,10 @@ public void testToggle() {
-        is(instanceOf(Preference.class)),
-        withKey("pref_bottomtoolbar")))
-        .perform(click());
+//    onData(allOf(
```
I would like this deleted, not just commented out

macgills commented:
@jhbiggs I am restarting the build, automated tests are notoriously flaky at the best of times, and I am running the tests locally, will report on my findings

macgills commented:
```
BUILD SUCCESSFUL in 3m 44s
```
that is for me locally so probably will pass on CI. Make the change I requested and this is ready for approval

macgills commented:
@jhbiggs wohoo build successful on ci, please now remove the commented out code

macgills commented:
also please make the target branch of this PR develop

jhbiggs commented:
Changed the target branch, deleted the test. 🤝

macgills commented on app/src/androidTestKiwix/java/org/kiwix/kiwixmobile/help/HelpActivityTest.java:
```diff
@@ -85,7 +85,7 @@ public void testHelpActivity() {
 
     onView(withText(context.getString(R.string.help_12))).check(matches(notNullValue()));
-    onView(withText(context.getString(R.string.help_12))).perform(click());
+  //  onView(withText(context.getString(R.string.help_12))).perform(click());
```
Uncomment this please

macgills replied:
I did not mean delete, there should be no changes to this test file because you did not alter HelpActivity.
Sorry I should have been more clear and said revert your changes

macgills commented on app/src/androidTestKiwix/java/org/kiwix/kiwixmobile/language/LanguageActivityTest.java:
```diff
@@ -80,7 +80,7 @@ public void testLanguageActivity() {
     // Open the Library
     openActionBarOverflowOrOptionsMenu(getInstrumentation().getTargetContext());
-    onView(withText("Get Content")).perform(click());
+   // onView(withText("Get Content")).perform(click());
```
Uncomment all changes in this file too

macgills replied:
revert also

macgills commented on app/src/androidTestKiwix/java/org/kiwix/kiwixmobile/tests/NetworkTest.java:
```diff
@@ -138,8 +138,8 @@ public void networkTest() {
 
-    onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.library_list))
-        .perform(click());
+//    onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.library_list))
```
Uncomment all changes in this too

macgills replied:
revert

macgills commented:
@jhbiggs please revert merging Unify UI icons in here, this is about issue #973 and that alone.
You also now have merge conflicts to resolve, I recommend merging develop in to your branch

macgills commented:
Hi @jhbiggs could you run the following command locally
```
 gradlew createKiwixDebugCoverageReport lintKiwixDebug jacocoTestKiwixDebugUnitTestReport assemble
```
We need to figure out how to allow people from forks to integrate with our CI, right now it has encrypted variables that could pose a security risk so it has blocked you.
If you can confirm for me that all of these checks pass I will be happy to approve and merge this PR, thank you for your contribution

macgills commented:
Hi @jhbiggs after some discussion we have decided the safest way for us to conduct our ci tests is by reopening this PR inside the repo. I have created a branch off of develop that I will now merge this PR into and will momentarily open a new PR that will run the CI checks
